src/client/windshaft.js

```diff
@@ -431,39 +432,109 @@ export default class Windshaft {
                 if (j == 0) {
                     throw new Error('Invalid MVT tile: first polygon ring MUST be external');
                 }
-                polygon.holes.push(polygon.flat.length / 2);
+                hole = true;
             }
+            let preClippedVertices = [];
             for (let k = 0; k < geom[j].length; k++) {
-                // TODO should additional clipping be done here?
-                let clipping = 0;
                 let x = geom[j][k].x;
                 let y = geom[j][k].y;
-
-                if (x > mvt_extent) {
-                    clipping = clipping | 1;
-                } else if (x < 0) {
-                    clipping = clipping | 2;
-                }
-                if (y > mvt_extent) {
-                    clipping = clipping | 4;
-                } else if (y < 0) {
-                    clipping = clipping | 8;
-                }
-                if (clipping) {
-                    polygon.clipped.push(polygon.flat.length);
-                    polygon.clippedType.push(clipping);
-                }
-                polygon.flat.push(2 * x / mvt_extent - 1);
-                polygon.flat.push(2 * (1 - y / mvt_extent) - 1);
+                x = 2 * x / mvt_extent - 1;
+                y = 2 * (1 - y / mvt_extent) - 1;
+                preClippedVertices.push([x, y]);
             }
+            this._clipPolygon(preClippedVertices, polygon, hole);
         }
         //if current polygon is not empty=> push it
-        if (polygon && polygon.flat.length > 0) {
+        if (polygon) {
             geometry.push(polygon);
         }
         featureGeometries.push(geometry);
     }
 
+    // Add polygon composed by preClippedVertices to the `polygon.flat` array
+    _clipPolygon(preClippedVertices, polygon, isHole) {
+        // Sutherland-Hodgman Algorithm to clip polygons to the tile
+        // https://www.cs.drexel.edu/~david/Classes/CS430/Lectures/L-05_Polygons.6.pdf
+        const clippingEdges = [
+            p => p[0] <= 1,
+            p => p[1] <= 1,
+            p => p[0] >= -1,
+            p => p[1] >= -1,
+        ];
+        const clippingEdgeIntersectFn = [
+            (a, b) => this._intersect(a, b, [1, -10], [1, 10]),
```

Comment:
Why 10?

Reply:
Anything larger than the extended tile would do?

Reply:
Yep (Anything larger than the extended tile should do it), I ran into some numeric problems when using `1`, the result was wrong by a very small amount, but just enough to mark it as no intersection which could create serious artifacts. The reason anything larger will work is that at that point we already know that there is an intersection point (because one point is outside the AABB and the other is inside).